stubs/SQLAlchemy/sqlalchemy/engine/interfaces.pyi

```diff
@@ -1,67 +1,149 @@
+    preexecute_autoincrement_sequences: bool
+    implicit_returning: bool
+    colspecs: dict[Any, Any]
+    supports_default_values: bool
+    supports_sequences: bool
+    sequences_optional: bool
+    supports_native_enum: bool
+    supports_native_boolean: bool
+    dbapi_exception_translation_map: dict[Any, Any]
+
     supports_statement_cache: bool
-    def create_connect_args(self, url) -> None: ...
+    @abstractmethod
+    def create_connect_args(self, url: URL) -> None: ...
+    @abstractmethod
     @classmethod
     def type_descriptor(cls, typeobj) -> None: ...
     def initialize(self, connection) -> None: ...
+    @abstractmethod
     def get_columns(self, connection, table_name, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def get_pk_constraint(self, connection, table_name, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def get_foreign_keys(self, connection, table_name, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def get_table_names(self, connection, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def get_temp_table_names(self, connection, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def get_view_names(self, connection, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def get_sequence_names(self, connection, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def get_temp_view_names(self, connection, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def get_view_definition(self, connection, view_name, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def get_indexes(self, connection, table_name, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def get_unique_constraints(self, connection, table_name, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def get_check_constraints(self, connection, table_name, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def get_table_comment(self, connection, table_name, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def normalize_name(self, name) -> None: ...
+    @abstractmethod
     def denormalize_name(self, name) -> None: ...
+    @abstractmethod
     def has_table(self, connection, table_name, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def has_index(self, connection, table_name, index_name, schema: Any | None = ...) -> None: ...
+    @abstractmethod
     def has_sequence(self, connection, sequence_name, schema: Any | None = ..., **kw) -> None: ...
+    @abstractmethod
     def do_begin(self, dbapi_connection) -> None: ...
+    @abstractmethod
     def do_rollback(self, dbapi_connection) -> None: ...
+    @abstractmethod
     def do_commit(self, dbapi_connection) -> None: ...
+    @abstractmethod
     def do_close(self, dbapi_connection) -> None: ...
+    @abstractmethod
```

Comment:
Have you checked that all subclasses override the abstract methods? If they don't, it's impossible to instantiate them.

Reply:
They don't, which is why we need quite a few `# type: ignore`s in the sub-classes. But please note that marking a method as abstract alone doesn't mean that sub-classes can't be instantiated. You also need to appropriate metaclass for that.

Reply:
> But please note that marking a method as abstract alone doesn't mean that sub-classes can't be instantiated. You also need to appropriate metaclass for that.

This is true at runtime, but not when type checking with mypy.

Reply:
:/ Well, I'll remove the @abstractmethod markers then and leave a comment.